geode-core/src/main/java/org/apache/geode/internal/cache/GemFireCacheImpl.java

```diff
@@ -375,6 +379,8 @@
 
   volatile boolean isClosing = false; // used in Stopper inner class
 
+  private final CountDownLatch isClosedLatch = new CountDownLatch(1);
```

Comment:
What happens when the cache is in reconnecting state. It did not looked like the latch is getting reset.